src/events/provider.py

```
@@ -39,27 +42,27 @@ def _on_service_account_requested(self, event: ServiceAccountRequestedEvent):
 
         if not self.charm.unit.is_leader():
             return
-        relation_id = event.relation.id
 
-        service_account = event.service_account
-        namespace = event.namespace
-        self.logger.debug(
-            f"Desired service account name: {service_account} in namespace: {namespace}"
-        )
+        if not event.service_account:
+            return
+
+        relation_id = event.relation.id
+        namespace, username = event.service_account.split(":")
```

**question**: should we validate the content and early return here if invalid, or is it done at the lib level?

A malformed service account will raise an exception here, effectively crashing the charm

---

**question, nitpick**: Additional nitpick, is `namespace:username` more common/obvious than `username:namespace`? Since we use `user:group` for setting file permissions, I tend to lean more toward the latter

---

I'm actually in favor of raising an exception here rather than silently returning early -- because I feel the early return from the event handler could make it difficult to detect something has actually gone wrong and this is something wrong in the implementation of the requirer charm (the requirer did not use the library properly). I'd be in favor of this validation if this was something derived from user configuration, and possibly also put the charm in blocked state to notify user that the config they have set is malformed. I'd also like to hear what @deusebio has to say on this.

---

Regarding the namespace-username combination, I think we should keep using `namespace:serviceaccount` because that's what is generally used in K8s (for eg, RBAC).